### DISCUSSION TITLE
Npm version used by renovate

### DISCUSSION THREAD

**Discussion by GuillaumeRahbari** [renovatebot/renovate#8553] [Request Help] [answered]
I wanted to know what version of npm renovate is using when running.

I have a repository where I use nvm to determine the node version I want to use. I'm currently using node version `14.15.4` that should come with the version `6.14.10` of npm. But when renovate update the `package-lock.json` file, there was a big change in it as it now uses the `lockfileVersion` `2` whereas with npm 6 it should be using the `lockfileVersion` 1.
I think this change happened because renovate is using the latest npm version (that is the version 7). But it is only a guess and I would like to know how it works and how we could configure it :)

**Comment by viceice** [comment, marked as answer]
Which renovate are you using? we recently fixed the npm v7 issue.

fixed on [v24.34.3](https://github.com/renovatebot/renovate/releases/tag/24.34.3) by #8506

So renovate will check the lockfile version and uses npm <7 if it's a v1 lockfile

**Reply by GuillaumeRahbari** [reply]
I'm using the renovate github app. Don't know which version it uses

**Reply by viceice** [reply]
is your project public? plrease use the retry checkbox to let renovate update the pr, maybe the pr was created before the fix was deployed

**Reply by viceice** [reply]
you can find the used version in the logs at https://app.renovatebot.com/dashboard

**Reply by GuillaumeRahbari** [reply]
We are using the version 24.38.6. So I guess we should have the fix now.

But if I understand well, renovate will use npm 6 or 7 depending on the version present in the lockfile ? And not depending of the npm version that comes from the nvmrc file (the version of npm that comes with node) ?

Is there a way to tell renovate to use the npm version that comes with the node version installed ?

**Reply by rarkins** [reply]
You can control the `npm` version  using `engines.npm` in `package.json`. There's no current way to say "use the default"

**Reply by GuillaumeRahbari** [reply]
Ok ! Thanks for your answers ! Help a lot.

**Reply by rarkins** [reply]
With the current approach you shouldn't find any example of incompatible lock files, e.g. caused by different versions of npm. If you do, then please discuss again so we can work out a way to fix that

**Reply by GuillaumeRahbari** [reply]
Ok ! Thanks again :)

**Comment by karfau** [comment]
I have the same issue and rebase/retry or recreating the PR using the Dependency Dashboard after it has been closed don't resolve the issue.
The repo is not public but I could provide read access for further analysis.
Is there any configuration required for the detection to work? (I didn't find anything when search the docs.)

**Reply by karfau** [reply]
from my package.json:
```json
  "engines": {
    "node": ">=12",
    "npm": ">=6"
  },
```

**Reply by karfau** [reply]
I just realized that there is both a `yarn.lock` file and a `package-lock.json` file in the repo. even though we are not using `yarn` in the repo, the update to `yarn.lock` looks like it could work. Just want to mention it in case it might be related.

**Reply by karfau** [reply]
In my case there was some weird issue on github action runner side, even though the lock file is updated to v2, suddenly npm ci works again.
I would still like to know how if there is anything that I can do to prevent this change from happening.

**Reply by karfau** [reply]
After changing the engine like this:
```json
  "engines": {
    "node": ">=12",
    "npm": "6"
  },
```
The Dependency Dashboard also lists an update to npm@7 and sticks with the existing lockfile version

**Reply by rarkins** [reply]
Needs reproduction in a public repo

**Comment by caiofbpa** [comment]
We're having this issue on a private repo at GitLab.

![Captura de Tela 2021-03-02 às 10 37 24](https://user-images.githubusercontent.com/507356/109656643-686e0c80-7b43-11eb-9848-873628b494d0.png)

We're using renovate's GitLab CI template:

![Captura de Tela 2021-03-02 às 10 40 15](https://user-images.githubusercontent.com/507356/109656925-b1be5c00-7b43-11eb-9766-3119ec3564af.png)

Not sure I can reproduce in a public repo.

**Reply by viceice** [reply]
Please use the dind version, that uses docker for running tools like npm and can use the required node version

```yaml
include:
  - project: 'renovate-bot/renovate-runner'
    file: '/templates/renovate-dind.gitlab-ci.yml'
```

**Reply by caiofbpa** [reply]
Just applied this new template and retried one dependency, got the same error "Unsupported engine"

**Reply by caiofbpa** [reply]
But now the npm version is correct, only the node version that's wrong:

![Captura de Tela 2021-03-02 às 11 10 20](https://user-images.githubusercontent.com/507356/109660734-f6e48d00-7b47-11eb-8539-f2a51b2905f0.png)

**Reply by caiofbpa** [reply]
I'll try removing node from engines because it's not so relevant for dependency installation and updates

**Reply by viceice** [reply]
please provide the debug logs, so we can see what's going wrong. renovate should use the correct version of node.

**Reply by caiofbpa** [reply]
Removing `node` from `engines` worked. Renovate runs with a different node version, but our CI tests still run in the correct node version so we're covered.

**Reply by viceice** [reply]
sure, but i want to know whats going wrong. so you should normally not required to remove the node dependency

**Comment by aabmass** [comment]
I'm seeing a similar issue on this public repo https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/237. The `package-lock.json` files use `lockfileVersion` 2, but renovate is still generating PRs with the old format. Any ideas?

I don't want to set the npm version in package.json `engines` because this is a library.

**Reply by viceice** [reply]
I think it's because you have node constraint in your package.json, so renovate will lot check the lockfile version.

**Reply by aabmass** [reply]
@viceice thanks for the reply. I need that for library users. I don't want to constrain npm either because users will see a warning if they aren't using npm7 locally. Any ideas how I can tell renovate to use `lockfileVersion` 2 without changing that?

**Reply by rarkins** [reply]
I don't see a reason why we shouldn't detect the need for npm 7 in this case. What do the logs indicate about it? Look for a log when the commit was generated.

**Reply by viceice** [reply]
`constraints` can be used to force renovate to use specific versions

https://docs.renovatebot.com/configuration-options/#constraints

**Reply by rarkins** [reply]
I've identified that this particular repo's behavior is a result of lerna being used. While for `npm`-only we have functionality to always `npm i -g` - meaning npm@7 if no constraint specified - for lerna we are only installing a new `npm` if there's a constraint present. I'll fix this

**Reply by rarkins** [reply]
Despite that fix, `lerna` keeps finding `npm@6` somehow. Looking on the lerna repo, I see more than one issue raised about lerna and npm@7 incompatibility, e.g.

https://github.com/lerna/lerna/issues/2822 and https://github.com/lerna/lerna/issues/2832

At this point the best recommendation I can give is to downgrade to `npm@6` if using lerna as I can't work out how to force `npm@7` and it's possibly a lerna limitation we can't change either.

**Reply by rarkins** [reply]
Hopeful solution: #8961

**Reply by aabmass** [reply]
@rarkins [I believe this is the log](https://app.renovatebot.com/dashboard#github/GoogleCloudPlatform/opentelemetry-operations-js/310779628) but I don't see anything about npm version. If you cannot see the link, [here is a gist of it](https://gist.github.com/aabmass/405d29e5c00ecd3613e55bbaf3113c85)

**Reply by aabmass** [reply]
@rarkins awesome, thank you for the quick PR! I'll let you know if it is fixed with that new version of renovate. I have noticed that lerna doesn't seem to use the same version of npm it is invoked (for example running `npx npm@^7 install` in that repo) but rather the global one.

**Reply by rarkins** [reply]
In our case it's not even using the npm in the path but instead doing it's own hunting for non and running from the directory where it comes preinstalled by node. We worked around that by forcing node 15 if Lerna is used and lockfileversion=2

**Reply by aabmass** [reply]
@rarkins I can confirm that PR fixed the issue, thank you again!
